cpp/tests/experimental/louvain_primitives.cu

```diff
@@ -0,0 +1,197 @@
+      num_verts, handle.get_stream());
+
+  copy_v_transform_reduce_key_aggregated_out_nbr(
+    handle,
+    graph_view,
+    thrust::make_zip_iterator(thrust::make_tuple(
+      old_cluster_sum_v.begin(), src_vertex_weight_v.begin(), src_cluster_weight_v.begin())),
+    cluster_v.begin(),
+    communities.begin(),
+    communities.end(),
+    community_in_weight_sums.begin(),
+    [total_edge_weight, resolution] __device__(
+      auto src, auto neighbor_cluster, auto new_cluster_sum, auto src_info, auto a_new) {
+      auto old_cluster_sum = thrust::get<0>(src_info);
+      auto k_k             = thrust::get<1>(src_info);
+      auto a_old           = thrust::get<2>(src_info);
+
+      weight_t delta_modularity = 2 * (((new_cluster_sum - old_cluster_sum) / total_edge_weight) -
+                                       resolution * (a_new * k_k - a_old * k_k + k_k * k_k) /
+                                         (total_edge_weight * total_edge_weight));
+
+      printf("compute delta_modularity (%d, %d, %g, %g, %g, %g, %g)\n",
+             src,
+             neighbor_cluster,
+             new_cluster_sum,
+             old_cluster_sum,
+             k_k,
+             a_old,
+             a_new);
+
+      return thrust::make_tuple(neighbor_cluster, delta_modularity);
+    },
+    [] __device__(auto p1, auto p2) {
+      printf("reducing (%d, %g) with (%d, %g)\n",
+             thrust::get<0>(p1),
+             thrust::get<1>(p1),
+             thrust::get<0>(p2),
+             thrust::get<1>(p2));
+
+      return (thrust::get<1>(p1) < thrust::get<1>(p2)) ? p1 : p2;
+    },
+    thrust::make_tuple(vertex_t{-1}, weight_t{std::numeric_limits<weight_t>::max()}),
+    // output_v.begin());
+    cugraph::experimental::get_dataframe_buffer_begin<thrust::tuple<vertex_t, weight_t>>(
+      output_buffer));
+
+  thrust::for_each(
+    thrust::device,
+    thrust::make_counting_iterator(0),
+    thrust::make_counting_iterator(1),
+    [b_output =
+       cugraph::experimental::get_dataframe_buffer_begin<thrust::tuple<vertex_t, weight_t>>(
+         output_buffer),
+     num_verts] __device__(auto) {
+      char comma[2] = {0, 0};
+      printf("output = ");
+      for (auto p = b_output; p != (b_output + num_verts); ++p) {
+        printf("%s(%d, %g)", comma, thrust::get<0>(*p), thrust::get<1>(*p));
+        comma[0] = ',';
+      }
+      printf("\n");
+    });
+  //
+  //  output_v contains (for each vertex) the pair (neighbor_cluster, delta_modularity)
+  //  for all neighboring clusters with the highest delta modularity
+  //
+  //  This can be used locally on each GPU to complete this portion of the algorithm
+  //
+}
+
+int main(int argc, char** argv)
+{
+  using vertex_t = int32_t;
+  using edge_t   = int32_t;
+  using weight_t = float;
+
+  weight_t resolution{1};
+
+  raft::handle_t handle{};
+
+  std::vector<vertex_t> src_h = {0, 0, 0, 0, 1, 1, 2, 2, 3, 4, 4, 4};
```

Comment:
shouldn't ```std::vector<vertex_t> src_h = {0, 0, 0, 0, 1, 1, 2, 2, 3, 4, 4, 4};``` be ```std::vector<vertex_t> src_h = {0, 0, 0, 0, 1, 1, 2, 2, 3, 3, 4, 4};```?

Reply:
Yes

Reply:
I got this ```output = (3, 0.263889),(1, 0.0972222),(1, 0.0972222),(3, 0.0972222),(3, 0.0972222)``` at the end and this sounds legitimate.

Reply:
With your bug fixes this seems to be working in a reasonable fashion now.  I did some experiments today to understand the exact behavior of the call.  I should be able to construct a complete implementation next week and try debugging that.